### PR TITLE
Spatial Elements Utilities and QOL

### PR DESCRIPTION
**Pull request by Gorgious56**
Following  https://github.com/IfcOpenShell/IfcOpenShell/issues/2004

This is a draft to see if I'm going in the right direction.

Right now this adds a button in the Ifc Aggregates panel when a building empty is selected, to add a new storey to it. I just stole code from the "CreateProject" operator to do it but it looks like it works well.

BSE_90

Todo list :

- [x] Add IfcBuilding to IfcSite button
- [x] Add IfcSite to IfcProject button
- [x] Use popup panel to let user choose a name for their new spatial structure ?
- [x] Display Related objects in the Aggregates panel
- [x] Add button to select the related object or relating object
- [ ] Add all these to the right click context menu in the outliner
- [ ] Find a good place for these operators in the 3D viewport
- [ ] Test suite
- [ ] UI Polish
- [ ] Bug hunting and fixing

### CODE REVIEW

File: src/blenderbim/blenderbim/bim/module/aggregate/ui.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 from bpy.types import Panel
 from blenderbim.bim.module.aggregate.data import AggregateData
 from blenderbim.bim.ifc import IfcStore
+from blenderbim.bim.module.root.data import IfcClassData
```
[the remark this answers is not in the document]
Comment (Gorgious56) [reply src/blenderbim/blenderbim/bim/module/aggregate/ui.py]: Alright I fixed it by simply adding the relevant method to the data class https://github.com/Gorgious56/IfcOpenShell/blob/e14b3b224e41fa3fbe6033cc3b1c9e0bdeedc3e1/src/blenderbim/blenderbim/bim/module/aggregate/data.py#L58-L62

**Comment by Moult** [comment]
> Am I using the whole operator / tool / core workflow correctly ?

Yes :)

> Is Ifc aggregates panel the right place for this button ?

Probably not. Maybe this is a candidate for the 3D workspace? This seems to me more like an authoring workflow. By the way, are you familiar with the usual BIM apps like ArchiCAD, Revit, Tekla? Would there be something there to inspire us?

Not too fussed though, it can move around. It's a good feature to have.

> When I select an IfcBuilding, shouldn't its relations to its stories also be displayed in the Ifc Aggregates panel ? Currently only the relating object higher in the hierarchy is displayed.

Good idea. Would you like to add that feature?

**Comment by Gorgious56** [comment]
@Moult Just pushed a proposition for displaying related objects in the aggregate panel, as well as an operator to unassign the relationship :


BES_60

I'm not pleased with the multi-box display in the aggregates panel but I don't really want to add yet another level of sub-panel for relating object and related objects. I'll explore an elegant way to display these informations.

I've renamed `get_aggregate` to `get_relating_object` since it seemed more descriptive of what the function was doing and it let me create a mirror `get_related_objects` but I'm not sure it was the right call and it's called from a lot of places in the code so I'd like to know if it was sensible : https://github.com/Gorgious56/IfcOpenShell//blob/spatial_elements_utilities/src/ifcopenshell-python/ifcopenshell/util/element.py#L181-L188

**Comment by Moult** [comment]
Yeah the multibox display isn't very slick. In the past I've simply used a label on each row, with each label using an icon to denote whether they are one type or another. See the implementation in voids.

Don't rename get_aggregate to get_relating_object. I made the same mistake in the past when I was trying to choose good variable names that mirrored the schema jargon. The problem is that there are a lot of different types of relating objects in IFC. There are also lots of different types of related objects. The appropriate antonym of get_aggregate might be get_part.

Also a word of warning that there may be many, many parts. A common example is clicking on a building and seeing 70 storeys. Or clicking on a storey and seeing 300 rooms (hospitals, anyone?). That's probably why I never bothered building it in the beginning. I don't know a good UI solution that isn't clunky.

**Comment by Moult** [comment]
Maybe a good solution to showing the parts in the UI is to simply say "N Parts Found" as a number, then provide a little "select icon" next to it so you can select all the corresponding objects visually.

**Comment by Gorgious56** [comment]
I reorganized the Aggregates panel a bit following your insights, looks more manageable now. Not sure about the icons but I couldn't find default ones that would convey the whole parts / aggregate thing. I'll maybe use the same things as in the openings panel.
Also added operators to select the parts or the relating element on the layout rows :

BSE_91

I'm thinking of adding a button that would open a popup panel to display the parts in a UIList that way it doesn't bother the user if they don't want to dig into it, but they still can access them directly from there.

**Comment by Gorgious56** [comment]
Extended the add_building_storey operator to allow creation of Site and Building :

BSE_92

**Comment by Moult** [comment]
Looking awesome! Feel free to merge!

**Pull request merged**